### ISSUE TITLE
ERRO[0000] 'overlay' is not supported over overlayfs  

### ISSUE THREAD

**Issue by cmoulliard** [projectatomic/buildah#867]
When I execute within a fedora:28 docker container this command

```
sudo buildah --debug images
```

I get this error

```
DEBU[0000] [graphdriver] trying provided driver "overlay" 
DEBU[0000] overlay: override_kernelcheck=true           
ERRO[0000] 'overlay' is not supported over overlayfs    
DEBU[0000] [graphdriver] trying provided driver "overlay" 
DEBU[0000] overlay: override_kernelcheck=true           
ERRO[0000] 'overlay' is not supported over overlayfs    
ERRO[0000] 'overlay' is not supported over overlayfs: backing file system is unsupported for this graph driver
'overlay' is not supported over overlayfs: backing file system is unsupported for this graph driver 
```

Here is the dockerfile used to create the buildah container on Macos

```
FROM fedora:28

RUN dnf install buildah sudo -y

#
# Create a user and group used to launch processes
# The user ID 1000 is the default for the first "regular" user on Fedora/RHEL,
# so there is a high chance that this ID will be equal to the current user
# making it easier to use volumes (no permission issues)
#
RUN groupadd -r default -g 1000 && useradd -u 1000 -r -g default -m -d /home/me -s /sbin/nologin -c "Default user" default

#
# Add user to sudo
#
RUN echo 'default:secret' | chpasswd
RUN echo '%default ALL=(ALL) ALL' >> /etc/sudoers

# Set the working directory to default' user home directory
WORKDIR /home/me
USER default
```

Docker info (command executed within macos's terminal)

```
docker info
Containers: 67
 Running: 11
 Paused: 0
 Stopped: 56
Images: 160
Server Version: 1.13.1
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: 
Swarm: inactive
Runtimes: docker-runc runc
Default Runtime: docker-runc
Init Binary: docker-init
containerd version:  (expected: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1)
runc version: N/A (expected: 9df8b306d01f59d3a8029be411de015b7304dd8f)
init version: N/A (expected: 949e6facb77383876aeff8a6944dde66b3089574)
Security Options:
 seccomp
  Profile: default
 selinux
Kernel Version: 3.10.0-693.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.954GiB
Name: minishift
ID: PWZQ:O5OI:AMTH:2IVK:E3FW:IX2G:BJ6X:JTWS:MW2N:YCY3:BDUI:LFAZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
 provider=xhyve
Experimental: false
Insecure Registries:
 172.30.0.0/16
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by rhatdan** [comment]
You need to volume mount storage on top of /var/lib/containers for this to work.
Or you could use a storage driver of something other then overlay, like vfs.

**Comment by cmoulliard** [comment]
I changed the storage to use "vfs" and I can dowload some images but when this error is next reported -> `failed to mount shm tmpfs`

Info 
```
[default@a0bd912f65f8 ~]$ sudo podman run -dt -e HTTPD_VAR_RUN=/var/run/httpd -e HTTPD_MAIN_CONF_D_PATH=/etc/httpd/conf.d                     -e HTTPD_MAIN_CONF_PATH=/etc/httpd/conf                     -e HTTPD_CONTAINER_SCRIPTS_PATH=/usr/share/container-scripts/httpd/                     registry.fedoraproject.org/f27/httpd /usr/bin/run-httpd
[sudo] password for default: 
Trying to pull registry.fedoraproject.org/f27/httpd...Getting image source signatures
Copying blob sha256:ff3dab903f926d26db009e0dcd575b8ccea7635b6b225835c35df765662707ec
 80.73 MB / 80.73 MB [=====================================================] 31s
Copying blob sha256:9347d6e9d8644ec22930b7dd78ba0cfcab8fd9a09d57545d4d74649ca929df0c
 7.30 MB / 7.30 MB [========================================================] 3s
Copying blob sha256:2fc5c44251d42d4e8bdc547a79e7ad1efffdf11f74ceea6a83e2fb6743aeff7d
 44.82 MB / 44.82 MB [===================================================] 2m14s
Copying config sha256:18f01f6f77ef941f5b31dd007d113620cbb1939ac33d1922edce7d9d2c9ebad7
 6.55 KB / 6.55 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
failed to mount shm tmpfs "/var/lib/containers/storage/vfs-containers/7a1b389dd27738244235fa6208f649637b0be651b74136c975f9ebd68fd43240/userdata/shm": operation not permitted

[default@a0bd912f65f8 ~]$ sudo ls -la /var/lib/containers/storage
[sudo] password for default: 
total 4
drwx------. 10 root root 149 Jul 17 19:41 .
drwx------.  3 root root  21 Jul 17 19:36 ..
drwxr-xr-x.  2 root root  27 Jul 17 19:41 libpod
drwx------.  2 root root   6 Jul 17 19:36 mounts
drwx------.  2 root root   6 Jul 17 19:36 overlay
-rw-------.  1 root root  64 Jul 18 04:25 storage.lock
drwx------.  2 root root   6 Jul 17 19:36 tmp
drwx------.  3 root root  17 Jul 17 19:42 vfs
drwx------.  3 root root 124 Jul 18 04:25 vfs-containers
drwx------.  4 root root 188 Jul 18 04:24 vfs-images
drwx------.  2 root root 384 Jul 18 04:25 vfs-layers
[default@a0bd912f65f8 ~]$ sudo ls -la /var/lib/containers/storage/vfs-containers
total 8
drwx------.  3 root root 124 Jul 18 04:25 .
drwx------. 10 root root 149 Jul 17 19:41 ..
drwx------.  3 root root  22 Jul 18 04:25 7a1b389dd27738244235fa6208f649637b0be651b74136c975f9ebd68fd43240
-rw-------.  1 root root 513 Jul 18 04:25 containers.json
-rw-------.  1 root root  64 Jul 18 04:25 containers.lock
```

**Comment by rhatdan** [comment]
Currently we only support these with SYS_ADMIN Capabilities.  Although @giuseppe Is close to having it run without any privs.  He has been looking at running podman while not root, but it should also run inside of a container.

**Comment by giuseppe** [comment]
https://github.com/containers/fuse-overlayfs can be used to mount overlay within an user namespace.  It requires Linux 4.18, that is at rc5 right now.  It works on Rawhide.  The biggest problem I see with running it inside of a container is that the container itself is running on top of overlay, so we need to check how that will work with whiteouts files.  There is a fallback mode to use files instead of devices in fuse-overlayfs, but should be checked on top of another overlay file system.

**Comment by rhatdan** [comment]
Closing this because I believe it has been answered.

**Comment by elgalu** [comment]
@giuseppe how can we do this by "use files instead of devices" via fuse-overlayfs ? I hit this `Error: 'overlay' is not supported over overlayfs`

**Comment by giuseppe** [comment]
@elgalu I am not sure I understand the question.  Would you like to use `fuse-overlayfs` on top of the overlay file system?

**Comment by elgalu** [comment]
Yes, on top of the overlay running inside a K8s unprivileged pod, is that possible?

**Comment by rhatdan** [comment]
Podman and Buildah need to be launched with at least SETUID and SETGID capabilities. Even if they run in a User Namespace, setting up the Namespace will require these CAPs in order to use more then one UID inside of the container.

**Comment by elgalu** [comment]
thanks @rhatdan , the pod has those capabilities:

```
Current: = cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_mknod,cap_audit_write,cap_setfcap+i
Bounding set =cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_mknod,cap_audit_write,cap_setfcap
Securebits: 00/0x0/1'b0
 secure-noroot: no (unlocked)
 secure-no-suid-fixup: no (unlocked)
 secure-keep-caps: no (unlocked)
uid=1000(nbuser)
gid=100(users)
groups=100(users),2000(???)
```

**Comment by rhatdan** [comment]
As long as you mount a volume on /var/lib/containers inside of you container then it should work.

You are not allowed to run overlayfs on top of a file system that is already overlay.

**Comment by elgalu** [comment]
We're not allowed to mount host's paths in this K8s setup, is there any other way beyond slower solutions like VFS ?

**Comment by rhatdan** [comment]
Set up fuse-overlayfs inside of the container, I think fuse-overlay will work on overlay.
@giuseppe Correct?

**Comment by giuseppe** [comment]
yes, fuse-overlayfs should work on top of overlay.

**Comment by liubin** [comment]
Thanks all, this worked for me to run buildah inside a container:

```
docker run --privileged -it -v /tmp/csivvv:/var/lib/containers --rm  centos:centos7 bash
# # mount |grep containers
/dev/nvme0n1p2 on /var/lib/containers type ext4 (rw,relatime,errors=remount-ro)

```
Just use `/var/lib/containers` as a volume.

And I wonder how to use fuse-overlayfs inside of the container to let it works, I have tried but failed:

```
[root@9237b68aec97 /]# mkdir lower upper work merged
[root@9237b68aec97 /]# 
[root@9237b68aec97 /]# fuse-overlayfs -o lowerdir=lower,upperdir=upper,workdir=work /var/lib/containers
[root@9237b68aec97 /]# mount
overlay on / type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay2/l/YEOKEZEWHFHL7WMM27BXFNQU6T:/var/lib/docker/overlay2/l/C4XFPXQ3OPSOKWEVRJ5B6FT5UW,upperdir=/var/lib/docker/overlay2/20f6a7c025dd2c6bdeebc32c00a561bb49119682c828c271432653f13f90c068/diff,workdir=/var/lib/docker/overlay2/20f6a7c025dd2c6bdeebc32c00a561bb49119682c828c271432653f13f90c068/work,xino=off)
fuse-overlayfs on /var/lib/containers type fuse.fuse-overlayfs (rw,nodev,noatime,user_id=0,group_id=0,default_permissions,allow_other)
[root@9237b68aec97 /]# buildah from --name myvolume --pull reg.docker.alibaba-inc.com/liubin/pouch-volume:latest
ERRO 'overlay' is not supported over <unknown> at "/var/lib/containers/storage/overlay" 
kernel does not support overlay fs: 'overlay' is not supported over <unknown> at "/var/lib/containers/storage/overlay": backing file system is unsupported for this graph driver
```

Any suggestion? @giuseppe @rhatdan

**Comment by giuseppe** [comment]
> `docker run --privileged -it -v /tmp/csivvv:/var/lib/containers --rm centos:centos7 bash`

you'll need at least to add the /dev/fuse device to the container (`--device /dev/fuse`)

**Comment by AntonOfTheWoods** [comment]
I was happy building all afternoon and have suddenly been hit by this. I didn't do anything to the system (Ubuntu 20.04 in WSL2) and it has been working fine for months. Now:

```
$ buildah images
kernel does not support overlay fs: 'overlay' is not supported over <unknown> at "/home/anton/.local/share/containers/storage/overlay": backing file system is unsupported for this graph driver
WARN failed to shutdown storage: "kernel does not support overlay fs: 'overlay' is not supported over extfs at \"/home/anton/.local/share/containers/storage/overlay\": backing file system is unsupported for this graph driver"
ERRO exit status 125
```

I have tried the `rm -rf` listed above, rebooted the WSL VM, rebooted the host. I have looked high and low and can't find anything. I didn't (intentionally) change anything, and now I can't even do a `buildah images`. I tried everything on https://chris.collins.is/podman-overlay-is-not-supported-over-xfs/ but it's not really relevant anyway. Is there some other possible reason for this happening?

**Comment by rhatdan** [comment]
Lookslike you have a new kernel that does not have overlay or the underlying  file system is not supported by overlay.

extfs

**Comment by AntonOfTheWoods** [comment]
> Lookslike you have a new kernel that does not have overlay or the underlying file system is not supported by overlay.
> 
> extfs

```
$ ll /mnt/c/Users/anton/linux/bzImage
-rw-r--r-- 1 anton anton 10079328 Oct 30 06:46 /mnt/c/Users/anton/linux/bzImage
```
It's the same since I installed it last October,
```
$ uname -ra
Linux dwin 5.10.0-rc1-next-20201029-microsoft-cbl #1 SMP Thu Oct 29 14:07:12 MST 2020 x86_64 x86_64 x86_64 GNU/Linux
```
It's a slightly customised version of the MS repo (with things like wireguard, etc.), and has been rock solid for almost 6mths now. I know I should update it, but that makes me pretty sure an update isn't the issue.

**Comment by SahilChaudhary25** [comment]
i am also getting this issue

Error: kernel does not support overlay fs: 'overlay' is not supported over extfs at "/var/lib/containers/storage/overlay": backing file system is unsupported for this graph driver

**Comment by AntonOfTheWoods** [comment]
> i am also getting this issue
> 
Did you just update?

**Comment by SahilChaudhary25** [comment]
I am using latest pod man in k8 container, not updating anything

**Comment by rhatdan** [comment]
You need to have a volume mounted on /var/lib/containers

**Comment by AntonOfTheWoods** [comment]
> You need to have a volume mounted on /var/lib/containers

Are you suggesting that there is something that changed with our systems, and not something to do with the update that I pulled in (via the Kubic apt repo)? It just seems very strange that something that requires "a volume mounted on /var/lib/containers" would work very well for months without one and then require one without anything having changed. If the answer is "go through the recent commit log, the changes are obvious" then so be it - I am just not sure that it is obvious to most users, and I wanted there to be a centralised place for people to find a resolution to this issue.

**Comment by rhatdan** [comment]
We are seeing issues in podman 3.1 and new Buildah with containers-storage on certain platforms not using fuse-overlayfs and falling back to overlay, which could be this issue.

**Comment by perezjasonr** [comment]
> After updating `buildah` this afternoon to version `100:1.20.0-3`, I (with an annoying delay that made me forget about the update) am now getting:
> 
> ```
> $ buildah images
> kernel does not support overlay fs: 'overlay' is not supported over <unknown> at "/home/anton/.local/share/containers/storage/overlay": backing file system is unsupported for this graph driver
> WARN failed to shutdown storage: "kernel does not support overlay fs: 'overlay' is not supported over extfs at \"/home/anton/.local/share/containers/storage/overlay\": backing file system is unsupported for this graph driver"
> ERRO exit status 125
> ```
> 
> It appears that installing `fuse-overlayfs` and uncommenting that in the `storage.conf` has made things work again. Is there some reason for this happening all of a sudden?

just updated buildah and now also getting this

**Comment by TomSweeneyRedHat** [comment]
@giuseppe thoughts?

**Comment by giuseppe** [comment]
Are you running inside of a container?

**Comment by AntonOfTheWoods** [comment]
> Are you running inside of a container?

A couple of us appear to be running inside WSL2, which is a full Hyper-V VM, not a container, though may have some subtleties.

**Comment by giuseppe** [comment]
Thanks, I am not familiar with WSL2.  On what file system is `/home/anton/.local/share/containers/storage/overlay` located?

Could you show me the output for `stat -f /home/anton/.local/share/containers/storage/overlay` ?

**Comment by AntonOfTheWoods** [comment]
> Could you show me the output for stat -f /home/anton/.local/share/containers/storage/overlay ?

As far as I understand, pretty standard `ext4`.

```
anton@dwin:~/dev$ stat -f /home/anton/.local/share/containers/storage/overlay
  File: "/home/anton/.local/share/containers/storage/overlay"
    ID: fedc9aa3bd65bc57 Namelen: 255     Type: ext2/ext3
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 65793553   Free: 32771167   Available: 29411628
Inodes: Total: 16777216   Free: 15079728
```

The key factor being that it was working, the upgrade happened, it stopped working, then (a couple of hours later) I installed `fuse-overlayfs` and everything started working again. "All's well that ends well" but I've had to pull out some hair with `buildah` for issues like this before, and it is an awesome tool, so feel that it would be heaps better to have a ready FAQ for others to not have to waste hours.

**Comment by findmyname666** [comment]
For those still guessing what they should do ...

You can mount any directory to `/var/lib/containers` where buildah running inside container will have write permissions. For those using selinux you need to do some tweaking. I do `chcon -Rt svirt_sandbox_file_t dir_name`.

You can create temp directory, mount it and delete it once you don't need it anymore:

```
DATA_DIR="$(mktemp -d -p '' 'var-lib-containers')"

podman run -ti --rm --mount type=bind,src="$DATA_DIR",dst=/var/lib/containers ...
```
